battleground_viewer/src/construct_render/construct_render.rs

```diff
@@ -580,287 +478,52 @@ impl ConstructRender {
-                use battleground_construct::util::cgmath::ToTranslation;
-                let p0_original = vec3(l.p0.0, l.p0.1, l.p0.2);
-                let p1_original = vec3(l.p1.0, l.p1.1, l.p1.2);
-                let p0_transformed = (transform * p0_original.to_h()).to_translation();
-                let p1_transformed = (transform * p1_original.to_h()).to_translation();
-                instanced.add_line(p0_transformed, p1_transformed, l.width, color);
-            }
-            _ => instanced.add(transform, color),
-        };
-    }
-
-    fn primitive_to_mesh(el: &display::primitives::Element) -> CpuMesh {
-        match el.primitive {
-            display::primitives::Primitive::Cuboid(cuboid) => {
-                let mut m = CpuMesh::cube();
-                // Returns an axis aligned unconnected cube mesh with positions in the range [-1..1] in all axes.
-                // So default box is not identity.
-                m.transform(&Mat4::from_nonuniform_scale(
-                    cuboid.length / 2.0,
-                    cuboid.width / 2.0,
-                    cuboid.height / 2.0,
-                ))
-                .unwrap();
-                m
             }
-            display::primitives::Primitive::Sphere(sphere) => {
-                let mut m = CpuMesh::sphere(32);
-                m.transform(&Mat4::from_scale(sphere.radius)).unwrap();
-                m
-            }
-            display::primitives::Primitive::Cylinder(cylinder) => {
-                let mut m = CpuMesh::cylinder(128);
-                m.transform(&Mat4::from_nonuniform_scale(
-                    cylinder.height,
-                    cylinder.radius,
-                    cylinder.radius,
-                ))
-                .unwrap();
-                m
-            }
-            display::primitives::Primitive::Cone(cone) => {
-                let mut m = CpuMesh::cone(128);
-                m.transform(&Mat4::from_nonuniform_scale(
-                    cone.height,
-                    cone.radius,
-                    cone.radius,
-                ))
-                .unwrap();
-                m
+            battleground_construct::display::primitives::Material::FenceMaterial(material) => {
+                self.fence_primitives.add_primitive(
+                    BatchProperties::None,
+                    el.primitive,
+                    element_transform,
+                    material.color.to_color(),
+                );
             }
-            display::primitives::Primitive::Line(_line) => CpuMesh::cylinder(4),
-            display::primitives::Primitive::Circle(circle) => {
-                let mut m = CpuMesh::circle(128);
-                m.transform(&Mat4::from_scale(circle.radius)).unwrap();
-                m
+            battleground_construct::display::primitives::Material::OverlayMaterial(material) => {
+                self.overlay_primitives.add_primitive(
+                    BatchProperties::None,
```

Comment:
It might be a nice property. I've done this for now (since it is the best we can do with the current render infra), but I might have an idea that works out even better given how this is an augmenting overlay.